File: 2698-find-the-punishment-number-of-an-integer/2698-find-the-punishment-number-of-an-integer.py
from re import fullmatch
class Solution:
    def valid(self, s: str, target: int, index: int, prev: int):
        if index >= len(s):
            if prev == target:
                return True
            return False

        for i in range(index + 1, len(s)+1):
            sa = s[index : i]
            if re.fullmatch(r"(?:0|(?:[1-9]+[0-9]*))", sa) and (curr := int(sa) + prev) <= target:
                if self.valid(s, target, i, curr):
                    return True
            elif fullmatch(r"0[0-9]+", sa):
                return False
        return False

    def punishmentNumber(self, n: int) -> int:
        return sum(num * num for num in filter(lambda i:((i*i)%9 <= 1) and self.valid(str(i*i), i, 0, 0), range(1, n+1)))

File: 2698-find-the-punishment-number-of-an-integer/2698-find-the-punishment-number-of-an-integer.c
#pragma GCC optimize("loop-unroll")
int valid(char *nums, int numsSize, int index, int curr, int target){
    if (index >= numsSize){ // If all characters from string is used
        if (curr == target){ // If using all characters we get target sum, success
            return 1;
        }
        return 0;
    }
    // Temporary string to store current partition of the string
    char currPart[8];
    memset(currPart, '\0', 8);
    // Check validity recursively for all indices
    for (int i = index; i < numsSize; i++){
        strncat(currPart, &nums[i], 1);
        if (valid(nums, numsSize, i+1, curr+atoi(currPart), target)){
            return 1;
        }
    }
    return 0;
}

int punishmentNumber(int n) {
    int ret = 0;
    for (int i = 1; i <= n; i++){
        int num = i*i;
        if ((num % 9) > 1){ // Early stopping
            continue;
        }
        // Convert the square into string
        int temp = num, size = floor(log10(num)) + 1;
        char nums[size];
        int index = size-1;
        while (temp){
            nums[index--] = (temp%10) + 48;
            temp /= 10;
        }
        // Recursive function to check validity
        if (valid(nums, size, 0, 0, i)){
            ret += num;
        }
    }
    return ret;
}